XamarinCommunityToolkit.UnitTests/ObjectModel/AsyncCommand_Tests.cs

```diff
@@ -0,0 +1,157 @@
+		[Theory]
+		[InlineData(true, null)]
+		[InlineData(true, true)]
+		[InlineData(false, 1)]
+		[InlineData(false, "string")]
+		public void CanExecute_Bool_Nullable_Parameter(bool canExecuteResult, object parameter)
+		{
+			// Arrange
+			var asyncCommand = new AsyncCommand<bool?>(i => Task.FromResult(true), i => true);
+
+			// Assert
+			Assert.Equal(canExecuteResult, asyncCommand.CanExecute(parameter));
+		}
+
+		[Fact]
+		public async Task Execute_MultipleExecution_Disabled()
+		{
+			// Arrange
+			var canExecuteChangedCount = 0;
+			var executeTcs = new TaskCompletionSource<bool>();
+			var asyncCommand = new AsyncCommand<bool?>(i => executeTcs.Task, i => true)
+			{
+				AllowMultipleExecution = false
+			};
+			asyncCommand.CanExecuteChanged += (sender, args) =>
+			{
+				canExecuteChangedCount++;
+			};
+
+			// Act
+			var executeTask = asyncCommand.ExecuteAsync(true);
+
+			// Assert
+			Assert.False(asyncCommand.CanExecute(true));
+			Assert.Equal(1, canExecuteChangedCount);
+			executeTcs.SetResult(true);
+			await executeTask;
+			Assert.True(asyncCommand.CanExecute(true));
+			Assert.Equal(2, canExecuteChangedCount);
+		}
+
+		[Fact]
+		public async Task Execute_MultipleExecution_Allowed()
+		{
+			// Arrange
+			var executionCount = 0;
+			var canExecuteChangedCount = 0;
+			var executeTcs = new TaskCompletionSource<bool>();
+			var asyncCommand = new AsyncCommand<bool?>(i =>
+			{
+				executionCount++;
+				return executeTcs.Task;
+			}, i => true)
+			{
+				AllowMultipleExecution = true
+			};
+			asyncCommand.CanExecuteChanged += (sender, args) =>
+			{
+				canExecuteChangedCount++;
+			};
+
+			// Act
+			var executeTask1 = asyncCommand.ExecuteAsync(true);
+			var executeTask2 = asyncCommand.ExecuteAsync(true);
+			asyncCommand.Execute(true);
+
+			// Assert
+			Assert.True(asyncCommand.CanExecute(true));
+			executeTcs.SetResult(true);
+			await Task.WhenAll(executeTask1, executeTask2);
+			Assert.True(asyncCommand.CanExecute(true));
+			Assert.Equal(3, executionCount);
+			Assert.Equal(0, canExecuteChangedCount);
+		}
+
+		[Fact]
+		public async Task Execute_OnException()
+		{
+			// Arrange
+			Exception onException = null;
+			var executeTcs = new TaskCompletionSource<bool>();
+			var asyncCommand = new AsyncCommand<bool?>(i => executeTcs.Task,
+				onException: exception =>
+				{
+					onException = exception;
+				});
+
+			// Act
+			var executeTask = asyncCommand.ExecuteAsync(null);
+			var exceptionResult = new Exception();
+			executeTcs.SetException(exceptionResult);
+
+			// Assert
+			await executeTask;
+			Assert.Equal(exceptionResult, onException);
+			Assert.True(asyncCommand.CanExecute(true));
+		}
+
+		[Fact]
+		public async Task Execute_NoParameter()
+		{
+			// Arrange
+			var executionCount = 0;
+			var executeTcs = new TaskCompletionSource<bool>();
+			var asyncCommand = new AsyncCommand(() =>
+			{
+				executionCount++;
+				return executeTcs.Task;
+			})
+			{
+				AllowMultipleExecution = true
```

Comment:
I am not sure that this is a proper fix.
Seems IsExecuting stay "True" after
```
executeTcs.SetResult(true);
await executeTcs.Task;
```

Is it some kind of deadlock or we should just add a small delay befoe assertion?
```
await Task.Delay(10);
```

Reply:
This is main problem with `async void`, know when it is complete. I'm not a big fan of the use of delays, the required delay will depend on the running machine and make flaky test result. I avoided the delay by releasing a semaphore on last `CanExecuteChanged` which must be raised on finally of last execute. What do you think about it?